### ISSUE TITLE
[BUG] trainer.evaluate() and topk_model.evaluate() do not give same metric (case of no sampled softmax)

### ISSUE THREAD

**Issue by murali-munna** [NVIDIA-Merlin/models#1148]
### Bug description

I am running experiments using sampled softmax: I am using new method (topk_encoder) to evaluate when am using sampled softmax. However, when am not using sampled softmax, I expected both the below methods to return the same metrics, however they do not.

### Steps/Code to reproduce bug

Original way of evaluation:
```
predict_last = mm.SequenceMaskLast(schema=seq_schema, target=target, transformer=xlnet_block)
eval_results = model_transformer.evaluate(
            valid_ds,
            batch_size=512,
            pre=predict_last,
            return_dict=True
        )
```

New way of evaluation:
```
target = train_ds_schema.select_by_tag(Tags.ITEM_ID).first
max_k = 10
topk_model = model_transformer.to_top_k_encoder(k=max_k)
topk_model.compile(run_eagerly=False)

loader = mm.Loader(valid_ds, batch_size=512)
eval_results = topk_model.evaluate(loader, return_dict=True, pre=predict_last)
```

### Expected behavior

In the case of no sampled softmax, expected both the above codes to provide the same result.

### Environment details

- Merlin version: 23.05 TF
- Tensorflow version (GPU?): 2.11.0+nv23.2

### Additional context

Let me know if you need more details.
